src/raft.rs

```diff
@@ -545,41 +546,159 @@ impl<T: Storage> Raft<T> {
         is_batched
     }
 
+    // Whether the given peer could use Follower Replication
+    fn use_delegate(&self, to: u64) -> bool {
+        if self.is_follower_replication_enabled() && self.is_leader() {
+            if self.groups.in_same_group(self.id, to) {
+                false
+            } else {
+                // If the target is not in the leader group and the group size > 1, use delegate
+                self.groups
+                    .get_members(to)
+                    .map_or(0, |members| members.len())
+                    > 1
+            }
+        } else {
+            false
+        }
+    }
+
+    #[inline]
+    fn is_follower_replication_enabled(&self) -> bool {
+        self.group_id == INVALID_ID
+    }
+
     /// Sends RPC, with entries to the given peer.
-    pub fn send_append(&mut self, to: u64, pr: &mut Progress) {
+    /// This can be called by the leader or a delegate. If the delegate call this, the field `to`
+    /// of produced message should be set to the leader id.
+    pub fn send_append(&mut self, to: u64, prs: &mut ProgressSet) {
+        if self.use_delegate(to) {
+            if let Some(gid) = self.groups.get_group_id(to) {
+                match self.delegated_msgs.get_mut(&gid) {
+                    Some(m) => {
+                        m.bcast_targets.push(to);
+                        if let Some(pr) = prs.get_mut(to) {
+                            // This just clean all the Inflights though there could be some MsgAppendResp on the way. And from
+                            // now the delegate will take over the flow control of the peer `to`.
+                            pr.become_delegated();
+                        }
+                        return;
+                    }
+                    None => {
+                        let members = self.groups.get_members(to).cloned().unwrap(); // this is safe because of the checking in `self.use_delegate`
+                        let delegate = match self.groups.get_delegate(gid) {
```

Comment:
I prefer `let delegate = self.pick_delegate(prs)`, which already contains `get_delegate` logic. And, its valid only if the `pr` is not paused.